File: inputs/2023/day-06/js/part2.js
import { readFileSync } from "fs"
const input = readFileSync("../input", "utf8").slice(0, -1);
const [time, distance] = input.split("\n").map(line => line.split(":")[1].trim()).map(item => item.replaceAll(" ", ""))


console.log(solve(time, distance))
function solve(times, distances) {
	const margin = []
	const alldist = calcTime(time)
	margin.push(alldist.filter(dist => dist > distances).length)
	console.log(margin.reduce((acc, val) => acc * val, 1))
}
function calcTime(time) {
	let distances = []
	for (let index = 0; index < time; index++) {
		let [speed, wait] = [index, index]
		const distance = speed * (time - wait)
		distances.push(distance)
	}
	return distances
}

File: inputs/2023/day-06/rust/src/bin/part2.rs
#[cfg(test)]
mod tests {
    use super::*;
    #[test]
    fn solving() {
        let input = "Time:      7  15   30
Distance:  9  40  200";
        let result = 71503;
        assert_eq!(solve(input), result);
    }
}

fn main() {
    let input = include_str!("../../../input");
    let result = solve(input);
    println!("{}", result);
}

fn solve(input: &str) -> i32 {
    let mut result = 0;
    let items = input.split_once("\n").unwrap();
    let distance = items
        .1
        .split_once(": ")
        .unwrap()
        .1
        .trim()
        .split(" ")
        .filter(|f| {
            if f == &"" {
                return false;
            }
            return true;
        })
        .collect::<Vec<&str>>();
    let mut distances = "".to_owned();
    for ele in distance {
        distances.push_str(ele)
    }
    let distancenum = distances.parse::<i64>().unwrap();
    let times = items
        .0
        .split_once(": ")
        .unwrap()
        .1
        .trim()
        .split(" ")
        .filter(|f| {
            if f == &"" {
                return false;
            }
            return true;
        })
        .collect::<Vec<&str>>();
    let mut timestr = "".to_owned();
    for ele in times {
        timestr.push_str(ele)
    }
    let timenum = timestr.parse::<i64>().unwrap();
    for item in 1..(timenum as usize) {
        let racetime = timenum - item as i64;
        let raceddistance = racetime * item as i64;
        if raceddistance > distancenum {
            result += 1;
        }
    }
    return result;
}
